src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs

```diff
@@ -115,21 +115,29 @@ public class Frame<TContext> : Frame
                                 await FireOnCompleted();
                             }
 
-                            _application.DisposeContext(context, _applicationException);
-                        }
-
-                        // If _requestAbort is set, the connection has already been closed.
-                        if (Volatile.Read(ref _requestAborted) == 0)
-                        {
-                            ResumeStreams();
-
-                            if (_keepAlive)
+                            try
                             {
-                                // Finish reading the request body in case the app did not.
-                                await messageBody.Consume();
+                                // If _requestAbort is set, the connection has already been closed.
+                                if (Volatile.Read(ref _requestAborted) == 0)
+                                {
+                                    ResumeStreams();
+
+                                    if (_keepAlive)
+                                    {
+                                        // Finish reading the request body in case the app did not.
+                                        await messageBody.Consume();
+                                    }
+                                }
+
+                                // ProduceEnd() must be called before _application.DisposeContext(), to ensure
+                                // HttpContext.Response.StatusCode is correctly set when
+                                // IHttpContextFactory.Dispose(HttpContext) is called.
+                                await ProduceEnd();
+                            }
+                            finally
```

Comment:
I think we can get rid of the finally above. I've wanted to clean up the control flow of this method for a while. It's not exactly buggy, but i's confusing because it's not immediately clear what we expect can throw.